src/systemd/sd-bus-vtable.h

```diff
@@ -90,7 +92,9 @@ struct sd_bus_vtable {
                 },                                                      \
         }
 
-#define SD_BUS_METHOD_WITH_OFFSET(_member, _signature, _result, _handler, _offset, _flags)   \
+#define SD_BUS_NAME(_name) #_name "\0"
```

Comment:
what is this used for?

Reply:
just to help the developer make sure the \0 is always appended, and see visually more names:
SD_BUS_NAME(firstname) SD_BUS_NAME(lastname) SD_BUS_NAME(city)

I have tried already this in a program I am writing, and it is easy to forget.

Mainly, it is to avoid a subtle problem: if the names are not enough and the last one is not terminated by \0 (that is, a double \0 at the end), and we skip length+1, then when we look for the next parameter we are no longer in the string memory.

Reply:
hmm, but the name is *very* generic. Also, can't we come up with a macro that takes a list of arguments and inserts the \0 between them and at the end automatically? i.e.

so that `SD_BUS_ARGSTR("foo", "bar", "baz")` gets resolved to `"foo\0bar\0baz\0\0`automatically?

Reply:
I don't know how to do such a macro, with variable number of arguments, but I can try.

As for the name being very generic, I noticed that but couldn't think of anything else less than 70 characters. I though immediately: now someone will use it convinced that it will automatic pass every test for path, method, and argument names.

Reply:
I got something from internet. Here is what it looks like, up to 1024 arguments:

`
#include <stdio.h>

#define FIRST(a, ...) a
#define SECOND(a, b, ...) b

#define EMPTY()

#define EVAL(...) EVAL1024(__VA_ARGS__)
#define EVAL1024(...) EVAL512(EVAL512(__VA_ARGS__))
#define EVAL512(...) EVAL256(EVAL256(__VA_ARGS__))
#define EVAL256(...) EVAL128(EVAL128(__VA_ARGS__))
#define EVAL128(...) EVAL64(EVAL64(__VA_ARGS__))
#define EVAL64(...) EVAL32(EVAL32(__VA_ARGS__))
#define EVAL32(...) EVAL16(EVAL16(__VA_ARGS__))
#define EVAL16(...) EVAL8(EVAL8(__VA_ARGS__))
#define EVAL8(...) EVAL4(EVAL4(__VA_ARGS__))
#define EVAL4(...) EVAL2(EVAL2(__VA_ARGS__))
#define EVAL2(...) EVAL1(EVAL1(__VA_ARGS__))
#define EVAL1(...) __VA_ARGS__

#define DEFER1(m) m EMPTY()
#define DEFER2(m) m EMPTY EMPTY()()
#define DEFER3(m) m EMPTY EMPTY EMPTY()()()
#define DEFER4(m) m EMPTY EMPTY EMPTY EMPTY()()()()

#define IS_PROBE(...) SECOND(__VA_ARGS__, 0)
#define PROBE() ~, 1

#define CAT(a,b) a ## b

#define NOT(x) IS_PROBE(CAT(_NOT_, x))
#define _NOT_0 PROBE()

#define BOOL(x) NOT(NOT(x))

#define IF_ELSE(condition) _IF_ELSE(BOOL(condition))
#define _IF_ELSE(condition) CAT(_IF_, condition)

#define _IF_1(...) __VA_ARGS__ _IF_1_ELSE
#define _IF_0(...)             _IF_0_ELSE

#define _IF_1_ELSE(...)
#define _IF_0_ELSE(...) __VA_ARGS__

#define HAS_ARGS(...) BOOL(FIRST(_END_OF_ARGUMENTS_ __VA_ARGS__)())
#define _END_OF_ARGUMENTS_() 0

#define MAP(m, first, ...)           \
  m(first)                           \
  IF_ELSE(HAS_ARGS(__VA_ARGS__))(    \
    DEFER2(_MAP)()(m, __VA_ARGS__)   \
  )(                                 \
    /* Do nothing, just terminate */ \
  )
#define _MAP() MAP

#define GREET(x) "Hello, " #x "!"
#define ZEROTERM(x) #x "\0"
#define TERM(x) #x "-"

#define SD_BUS_ARGSTR(...)	EVAL(MAP(TERM, __VA_ARGS__));

void main(void) {
	char *s = SD_BUS_ARGSTR(a, b, c, d, e, f, g, h, i, j);
	printf("%s\n", s);
}

`

it comes from http://jhnet.co.uk/articles/cpp_magic
I don't know what kind of copyright can be associated with it, but I am asking.
At most I will have to rewrite it, but it will work that way

Reply:
yeah, looks a bit over the top, indeed. maybe do the simple macro then, but plz name it properl